Database

hoslack opened the pull request:
- Created an express server and added mongoose to access an online  database 
-  Made a route for reports and controller to add report to the database
Closes #20

hoslack committed:
Created server and database
Created an express server and added mongoose to access an online  database
Made a route for reports and controller to add report to the database
relates #20

Karyum commented:
love this priority 3 label on the PR :laughing: it's like saying oh it's not really important to review this right now :rofl:

Karyum reviewed (changes requested):
ohh no i done screwed up...
I should have noticed that you have more than 1 `package.json` before....

Cause you don't need 3 of them to run the server react and the app itself, 1 is enough it's just gonna be a mess of a file structure loads of `package-lock.json` and very confusing

apologies for not catching this before, and although we might get merge conflicts i need you guys to move everything ( server, react ) into 1 main `package.json` in the root directory of you project.

Karyum commented:
@hoslack  check my review then assign @MynahMarie @MarlenAw to check it out and then reassign me

hoslack commented:
@Karyum Please merge the first PR first before I make my changes, is that possible?

Karyum commented:
@hoslack yeah but why do you want me to merge this one ?
the best case senario here it to merge @MynahMarie PR #21 and then you would pull the new master and do the changes. don't worry i'll be here as soon as you make the new PR to check this. cause this PR will block your project progress, make sure that @MynahMarie @MarlenAw review the PR cause this is a big change

hoslack commented:
@Karyum  Kindly let us do the refactoring in the next PR, the conflict that's coming ahead is like lightning!

Karyum commented:
@hoslack 
we could avoid this conflict by doing what i told you, wait untill #21 gets merged and then do the changes on your branch and barabing barabong, if there is a better way you think this can happen then please explain it.

hoslack commented:
Please help me understand what I should do, maybe I don't get it. Thanks

Karyum commented:
as i said [here](https://github.com/FACN3/easyRep/pull/22#issuecomment-364741203) wait until #21 gets merged which from the looks of it gonna be soon its just a matter of commiting into the branch and then pull the new master and do the change that i requested

was that clear?

hoslack commented:
That's crystal clear. I just have to wait. Great!

hoslack committed:
Pulled changes from master
relates #20

hoslack committed:
removed readme created by create-react-app
relates #20

hoslack commented:
Hello, @Karyum Allow us to use the 2 package.json files in development and later we will run a build script to minimize what we put in production. That's the easiest way to use create-react-app. We have tried using one package file but it still asking for the other one inside the client!

Karyum commented:
i don't understand at what point create-react-app tell you, that you need another package.json so you get it to work, the few times i have used create react app i didn't have to do this, currently if i want to use a project to learn how to use create react app  i would use @matthewdking 's one https://github.com/facn2/path-18-app and he has only 1 package.json there.

But if you insist that you want to do it this way, and your not sure how to do it other ways then i'll be waiting for @MynahMarie and @MarlenAw 's approval on this so i can start code reviewing

MynahMarie commented on models/Reports.js:
```diff
@@ -0,0 +1,12 @@
+const mongoose = require('mongoose');
+const { Schema } = mongoose;
+
+const reportsShema = new Schema({
```
Please correct the little typo there... (reportsShema should be reportsSchema)

hoslack commented:
Thanks Mario or @Karyum . We are not Gurus yet but we are heading there.

MynahMarie commented on routes/reportRoutes.js:
```diff
@@ -0,0 +1,7 @@
+const createReport = require('../controllers/createReport');
+
+//add report to the database
+router.post('/create', createReport);
```
Would it be ok to change the endpoint to something like 'submit' instead of 'create'? because we will only send the report to the database if the user actually reviews and sends it right? (correct me if I'm missing something...)

hoslack replied:
@MynahMarie, I think submit is more for a form in the front end, create means making a new collection in the database. This is for simple understanding for someone who might want to use the api even for a different project

MynahMarie replied:
@hoslack ok then!

Karyum replied:
so this is `/create` but what is sit creating so i would say add to the endpoint the thing you are creating cause it is not obvious

MynahMarie commented on app.js:
```diff
@@ -0,0 +1,22 @@
+
+const port = 5000 || process.env.PORT;
+app.listen(port, () => {
+  console.log(`App listening on http://localhost:${port}`);
```
Can we add a URL variable instead of hardcoding localhost? (Im just thinking of when we will deploy... I know its far ahead but it's such a small detail, maybe we can do it now?) I know its just a tiny detail...

hoslack replied:
Done @MynahMarie

MynahMarie reviewed (changes requested):
Just little details @hoslack otherwise it looks really good!

hoslack committed:
Corrected typo in report model and used an env variable for a host
relates #20

Karyum commented on app.js:
```diff
@@ -0,0 +1,23 @@
+
+const port = 5000 || process.env.PORT;
+const host = 'localhost' || process.env.HOST;
+app.listen(port, () => {
```
Hmm don't you want to run your server when you database actually connected ?
so `mongoose.connect` can have event listeners, for example on error and on connected.

check [this out](http://theholmesoffice.com/mongoose-connection-best-practice/#managingthemongooseconnection)
so i would highly suggest to add the error event listener and the `connected` listener for when the database has connected and then you would want your server to connect after the db

hoslack replied:
Great, done!!

Karyum reviewed (changes requested):
Schemas might need more validation but i guess this is fine for now. :shipit:

hoslack commented:
@Karyum Already changed. Great changes!

hoslack committed:
Added error checking to mongoose
The server now only runs after the mongoose is connected otherwise it throws an error
changed endpoint from create to create_report
relates #20
